Blog CML self-hosted runners on demand with GPUs (#1656)
iterative/dvc.org: content/blog/2020-08-05-cml-self-hosted-runners-on-demand-with-gpus.md

runners: runners that you manage, but are accessible to your CI/CD system for
executing jobs. It could be an EC2 instance or the GPU under your desk. In our
[recently-released project](https://dvc.org/blog/cml-release), Continuous Machine Learning (CML), our Docker image acts as a thin wrapper
over GitLab and GitHub runners, adding some extra capabilities.

Here are some benefits of using CML as a self-hosted runner:

1.  Easy to use. Working the same way for both GitLab and GitHub.

2.  Get out of dependency hell. We tend to install packages (on top of packages,
    on top of packages…) while we‘re experimenting with models. In ML in
    particular, we can be dependent on drivers AND libraries, and sometimes
    precise versions of them (CUDA and TensorFlow, anyone?). Your CI workflow
    will install all the dependencies in the containerised runner leaving your
    machine always clean.

3.  Security. If your repo is public your runners could be accesed by anyone
    that could add
    [scripts that exploits your machine](https://docs.GitHub.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories).

>     With the containerised runner you are restrifying the access to your real

Reviewer: erm it's possible to break out of a container since the docker daemon has root privileges. Sort of the entire reason Singularity exists (maybe we could support Singularity?) related to https://github.com/iterative/cml/issues/191

I'd recommend leaving out this point (3) for now?

If we're keeping it; `restrifying` -> `restricting`

Author: @casperdcl I did not know about the Singularity project, is this [one](https://sylabs.io/singularity/)? About breaking out of the container... one thing to remember is that we are not mounting the docker socket, without it Im not saying that it's not totally impossible... but... I can't figure out right now how to exploit the host in that conditions. Could you please point me to some docs or findings?

Reviewer: yes that's the project.

The exploits are constantly changing and I think docker actively removes posts about it to ensure it isn't trivial; so my guess is most currently working exploit writeups are on the dark web.

I've worked with orgs which insist on singularity over docker purely to block potential privilege escalation.

![](/uploads/images/2020-08-05/registered-cml-runner-GitHub.png)

![](/uploads/images/2020-08-05/registered-cml-runner-GitLab.png)

### 3) Setup your GitHub Actions or GitLab workflow yaml file to use the runner and commit your changes.

GitLab

```yaml
train:
  tags:
    - cml
    - gpu

  script:
    - echo 'Hi from CML!' >> report.md
    - cml-send-comment report.md
```

GitHub

```yaml
name: train-my-model

on: [push]

jobs:
  train:
    runs-on: [self-hosted, cml, gpu]

    steps:
      - uses: actions/checkout@v2

      - name: cml_run
        run: |
          echo 'Hi from CML!' >> report.md
          cml-send-comment report.md
```

Congrats! At this point you have done all the steps to have your GPUs up and
running with CML.

# Limitations and future directions

There are still some limitations to be solved at this stage: 

 - GitHub Actions [can’t run a workflow longer than 72 hours](https://docs.GitHub.com/en/actions/getting-started-with-GitHub-actions/about-GitHub-actions#usage-limits).

 - Self-hosted runners [don’t behave well when they disconnect from the repo](https://GitLab.com/GitLab-org/GitLab/-/issues/229851#note_390371734),

> limiting the possibilities with preemptible instances (also known as spot instances).

Reviewer: link to issues where these are being addressed?

P.S. easy to get around the 72hr limit by having the runner spin off long running jobs in separate processed and exiting quickly. The background jobs can just update/add comment(s) to the relevant PR whenever they complete.

Author: @casperdcl thanks for the suggestion! I have tried your idea, in fact, I have some on top of the table, however... what I really dislike if the workflow fail per se. Sending you a message etc, and having to restart it to pass it having of course to use a cache system to not start training once the workflow restarts. One key thing to remember is that the check should be green always.

---
title: CML self-hosted runners on demand with GPUs
date: 2020-08-05
description: |

>   Bring reproducibility and GPUs to your ML projects in Github or Gitlab pipelines with CML self-hosted runners.

Reviewer 2: Throughout, let's use the format "GitHub" and "GitLab" instead of "Github" and "Gitlab".

Also for short description here's a suggestion:
``` Use your own GPUs with GitHub Actions & GitLab CI for automated machine learning.```

descriptionLong: |

>   When we train our models we usually need special hardware like big memory or GPUs. 

Reviewer 2: Let's shorten this: 

```Training models often requires special hardware, like extra memory or GPUs. How can we provide a CI/CD pipeline with this hardware? Find out how to setup your own self-hosted runners on-demand with GPUs for fast training.```

  How can we provide our CI/CD pipeline with such hardware?
  Even more, how could we ensure the reproducibility of the experiment? 
  Find out how to set up your own self-hosted runners on demand with GPUs for fast training and ensuring reproducibility in this blog post!
picture: 2020-08-05/header.png
author: david_g_ortega
commentsUrl: https://discuss.dvc.org/t/cml-self-hosted-runners-on-demand-with-gpus/462
tags:
  - CML
  - Continuous Machine Learning
  - GPUs
  - Self-hosted Runner
  - Reproducibility
  - Tutorial
---

> When setting your CI/CD workflow for your machine learning (ML) project you

Reviewer 2: Proposing a minor revision for grammar:

```
When creating your CI/CD workflow for a machine learning (ML) project,
you might find that by default, neither GitHub Actions nor GitLab CI provides 
the computing capabilities you need- like GPUs, high memory instances, 
or multiple cores. 

To overcome this hardware hurdle....
```

What do you think?

might find that not Github nor Gitlab provides certain computing capabilities
like GPU, high memory, big processors... capabilities that really matter in some
machine learning (ML) problems.

To overcome this hardware hurdle, one practical approach is to use self-hosted
runners: runners that you manage, but are accessible to your CI/CD system for
executing jobs. It could be an EC2 instance or the GPU under your desk. In our

> recently-released project, Continuous Machine Learning (CML), we’ve introduced a

Reviewer 2: let's link to CML release blog post!

tool called cloud-runner. It's a docker container that acts as a thin wrapper
over GitLab and GitHub runners, adding some extra capabilities.

> Benefits of using CML cloud runner:

Reviewer 2: For grammar, I'd rewrite:

`Here are some benefits of using CML cloud-runner:`

Reviewer 2: Also- are we still calling it cloud-runner since the cloud-runner docker image has been merged into base cml?

Author: > Also- are we still calling it cloud-runner since the cloud-runner docker image has been merged into base cml?

I thought the same, however, it could be like and internal name? I mean something that we know it exists...

Author: ```Here are some benefits of using CML as a self-hosted runner:```

unregisters from your repo.

`RUNNER_LABELS` a comma delimited list of labels that we are setting in our
workflow that the jobs will wait for.

`RUNNER_REPO` is the url of your Gitlab or Github repo. repo_token is the
personal token generated for your Github or Gitlab repo. Note that for Github
you must check `workflow` along with `repo`.

If everything went fine we should see a runner registered in our repo.

![](/uploads/images/2020-08-05/registered-cml-runner-github.png)

![](/uploads/images/2020-08-05/registered-cml-runner-gitlab.png)

### 3) Setup your Github Actions or Gitlab workflow yaml file to use the runner and commit your changes.

Gitlab

```yaml
train:
  tags:
    - cml
    - gpu

  script:
    - echo 'Hi from CML!' >> report.md
    - cml-send-comment report.md
```

Github

```yaml
name: train-my-model

on: [push]

jobs:
  train:
    runs-on: [self-hosted, cml, gpu]

    steps:
      - uses: actions/checkout@v2

      - name: cml_run
        run: |
          echo 'Hi from CML!' >> report.md
          cml-send-comment report.md
```

Congrats! At this point you have done all the steps to have your GPUs up and
running with CML.

# Limitations and future directions

> There are still some limitations to be solved at this stage: Github Actions

Reviewer 2: Let's use a bulleted list here.

``` There are still some limitations to be solved at this stage:
- GitHub Actions can't....
- Self-hosted runners don't behave well...
```

---
title: CML self-hosted runners on demand with GPUs
date: 2020-08-05
description: |
  Use your own GPUs with GitHub Actions & GitLab CI for automated machine learning.
descriptionLong: |

>   Training models often requires special hardware, like extra memory or GPUs. How can we provide a CI/CD pipeline with this hardware? Find out how to setup your own self-hosted runners on-demand with GPUs for fast training.

Reviewer 3: 💅 under <80 please

Reviewer 3: still the same problem?

Reviewer 3: @DavidGOrtega PTAL - a minor one

Author: @shcheklein 
```
Use your own GPUs with GitHub Actions & GitLab for continuous machine learning.
```

Its 79!

Reviewer 3: https://github.com/iterative/dvc.org/pull/1656/files#diff-348c0fb37e16bfa16cd2c3f0bee054ebR7 - It's def > 80 for me. Am I missing something?

Author: Are you speaking about descriptionLong? Sorry I though it was description...

Author: so why use the description vs descriptionLong? How long should be description?

Reviewer 3: yes, sorry. it's all about formatting - not the text length per se. Unfortunately prettier does not handle frontmatter automatically.

Author: So should I repeat the same text in the description but in the descriptionLong?
Should we add a ticket to include that i docs?

Reviewer 3: It's about formatting - just wrap the text properly with new lines. Does not matter how actually text long is - for both description and descriptionLong - they can be arbitrary long. (obviously, usually, descriptionLong is longer :) ).

    is perfectly reproducible by anyone in the future.

## Hands on GPU Self-hosted runners 101

### 1) Install nvidia drivers and nvidia-docker in your machine (ubuntu 18.04)

```dvc
$ curl -s -L https://nvidia.GitHub.io/nvidia-docker/gpgkey | sudo apt-key add - && \
  curl -s -L https://nvidia.GitHub.io/nvidia-docker/ubuntu18.04/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list && \
  sudo apt update && sudo apt install -y ubuntu-drivers-common  && \
  sudo ubuntu-drivers autoinstall  && \
  sudo apt install -y nvidia-container-toolkit && \
  sudo systemctl restart docker
```

You can test that your gpus are up and running with the following command:

```dvc
$ docker run --gpus all dvcorg/cml-py3 nvidia-smi
```

We should see something like this:
![](/uploads/images/2020-08-05/nvidia-smi-output.png)

### 2) Start your self-hosted runner

With CML docker images launching your own self-hosted runner is very easy. They
comes in two flavours, depending if you need python 2 or 3:

- dvcorg/cml:latest
- dvcorg/cml-py3:latest

These images have CML and DVC preinstalled (among other perks), plus CUDA

> drivers and Python 2.7. The cml-py3 image comes with Python 3.X as well. That's

Reviewer 3: curious - why do we support Python 2 still?

Author: many projects in ML space only supports python 2

Reviewer 3: hmm ... any stats on this?

Author: No

Reviewer 3: does it mean that with the default names like `dvcorg/cml` we actually provide Python 2 image?

Author: Not a python version per se. Simply we do not provide one, its de version that comes bundled with ubuntu 18.04 which is the one used in GH

Reviewer 3: ah, I see ... that's really sad to see that things move so slow, even for Ubuntu and GH

Author: true

Reviewer: > comes with Python 3.X as well

That's weird though. Implies that `cml-py3` bloatedly has both py2 and py3 packages pre-installed. Hopefully that is not the case and this should be reworded.

Reviewer: ```suggestion
drivers and Python 2.7 (use the cml-py3 image for Python 3.X). That's
```

Author: actually that sentence is very redundant. Its being explained above. im removing it

all. You can clone these images and add your own dependencies to better mimic
your own production environment.

```dvc
$ docker run --name myrunner -d --gpus all \
    -e RUNNER_IDLE_TIMEOUT=1800 \
    -e RUNNER_LABELS=cml,gpu \
    -e RUNNER_REPO=$my_repo_url \
    -e repo_token=$my_repo_token \
    dvcorg/cml-py3
```

where:

`RUNNER_IDLE_TIMEOUT` is the time in seconds that the runner is going to be idle
at most waiting for jobs to come, if no one comes the runner shuts down and
unregisters from your repo.

`RUNNER_LABELS` a comma delimited list of labels that we are setting in our
workflow that the jobs will wait for.

`RUNNER_REPO` is the url of your GitLab or GitHub repo. repo_token is the
personal token generated for your GitHub or GitLab repo. Note that for GitHub
you must check `workflow` along with `repo`.

If everything went fine we should see a runner registered in our repo.

> ![](/uploads/images/2020-08-05/registered-cml-runner-GitHub.png)

Reviewer 3: that's what I see:

<img width="712" alt="Screen Shot 2020-08-07 at 10 25 41 AM" src="https://user-images.githubusercontent.com/3659196/89671852-8e2a8000-d898-11ea-9894-e826aeca49ee.png">

Author: @shcheklein this is what I see

![image](https://user-images.githubusercontent.com/414967/89672150-898bb900-d8e4-11ea-929c-7177f763d2f6.png)

Author: ah! 
got it!

Author: its case sensitive @shcheklein, my bad not having tested this on linux vs macos

    - cml-send-comment report.md
```

GitHub

```yaml
name: train-my-model

on: [push]

jobs:
  train:
    runs-on: [self-hosted, cml, gpu]

    steps:
      - uses: actions/checkout@v2

      - name: cml_run
        run: |
          echo 'Hi from CML!' >> report.md
          cml-send-comment report.md
```

Congrats! At this point you have done all the steps to have your GPUs up and
running with CML.

# Limitations and future directions

There are still some limitations to be solved at this stage:

- GitHub Actions
  [can’t run a workflow longer than 72 hours](https://docs.GitHub.com/en/actions/getting-started-with-GitHub-actions/about-GitHub-actions#usage-limits).

- Self-hosted runners
  [don’t behave well when they disconnect from the repo](https://GitLab.com/GitLab-org/GitLab/-/issues/229851#note_390371734),
  limiting the possibilities with preemptible instances (also known as spot
  instances).

We’re working on both these issues
([#161](https://github.com/iterative/cml/issues/161)
[#174](https://github.com/iterative/cml/issues/174)
[#208](https://github.com/iterative/cml/issues/208)) both in terms of CML and
DVC capabilities. So keep watching this space for updates!

> 

Reviewer 3: Let's insert `<hr />` - it'll look better

  [can’t run a workflow longer than 72 hours](https://docs.GitHub.com/en/actions/getting-started-with-GitHub-actions/about-GitHub-actions#usage-limits).

- Self-hosted runners
  [don’t behave well when they disconnect from the repo](https://GitLab.com/GitLab-org/GitLab/-/issues/229851#note_390371734),
  limiting the possibilities with preemptible instances (also known as spot
  instances).

We’re working on both these issues
([#161](https://github.com/iterative/cml/issues/161)
[#174](https://github.com/iterative/cml/issues/174)
[#208](https://github.com/iterative/cml/issues/208)) both in terms of CML and
DVC capabilities. So keep watching this space for updates!

We started CML to help teams deal with the complexity of ML more effectively-
continuous integration is a proven approach to keeping projects agile even as
the team size, number of experiments, and number of dependencies increase.
Treating experiments like potential new features in a software project opens up
many possibilities for improving our engineering practices. We’re looking
forward to an era when ML experiments can be created, logged, and merged into
production-ready code in minutes, not days or weeks.

> <hr />

Reviewer 3: I meant `<hr>` before the last paragraph ... kinda conclusion one

---
title: CML self-hosted runners on demand with GPUs
date: 2020-08-05
description: |

>   A tutorial on how to setup your own self-hosted runners with GPUs using CML cloud-runners in your CML workflow in Github or Gitlab.

Reviewer 4: ```suggestion
  A tutorial on how to setup your own self-hosted runners with GPUs using CML cloud-runners on your CML workflow in Github or Gitlab.
```

>   A tutorial on how to setup your own self-hosted runners with GPUs using CML cloud-runners in your CML workflow in Github or Gitlab.

Reviewer 3: please fit it under 80

+ add desriptionLong, see in some other posts how it's done. descriptionLong support basic Markdown and appears in the post before the title image. It's a summary usually, or can serve as an intro

Reviewer 3: I would say in the descriptionLong let's try to expand a bit on what problem and how will people learn to solve it in the blog post.

picture: 2020-08-05/header.png
author: david_g_ortega

> commentsUrl: https://discuss.dvc.org/t/cml-self-hosted-runners-on-demand-with-gpus/??

Reviewer 3: we can go and create the topic I guess //cc @elleobrien

Reviewer 2: Yep I'll open it now

Reviewer 2: https://discuss.dvc.org/t/cml-self-hosted-runners-on-demand-with-gpus/462

Reviewer 3: 👍

tags:
  - CML
  - GPUs

>   - Self-hosted runner

Reviewer 3: runner -> Runner

Reviewer 2: I think lowercase might be right here- I usually don't capitalize all words of tags, just the first

Reviewer 3: hmm, those that I checked are capitalized ... obviously not a big deal though :)

Reviewer 3: https://github.com/iterative/dvc.org/issues/1657#issuecomment-668793001 - the list of tags

  - Continuous Machine Learning
  - Github
  - Gitlab

>   - Technical debt

Reviewer 3: let's use Reproducibility instead

>   - Tutorial

Reviewer 4: Probably too many tags (and too specific ones). You can check https://dvc.org/blog/tags/{tag} to see if there's any content there already.

Not sure what our blog post tagging guidelines are actually.

Reviewer 4: BTW @elleobrien https://dvc.org/blog/tags/CML has no content, may want to add that one to some prior blog posts? 🙂

Reviewer 2: @jorgeorpinel good idea!

---

> # CML self-hosted runners on demand with GPUs

Reviewer 3: I think title is not needed - it generates one from the frontmatter markup above

> Normally, CI/CD workflows are set up as a collection of jobs that are picked and

Reviewer 3: The whole first part is very long and monotonous I would say. At the very least let's add links (like self-hosted runners), use subtitles, simplify a bit, use inline code when it's needed, etc.

    eventual production environment. Using a CI system that codifies

3.  Always know what you did. One of the biggest technical debts in the ML space
    is reproducibility. A few weeks post-experiment, we often discover that
    trying to put your model back in shape is a pain. Looking at our repo, it’s
    not obvious what data or training infrastructure or dependencies went into a
    given result. Having a record that an experiment happened isn’t the same as
    being able to recreate it. But just moving your ML experiments into a CI/CD
    system constrains the environment and hardware used for your experiment-the
    workflow is perfectly isolated by the containerised runner and perfectly
    reproducible by yourself in the future or even other peers.

Benefits of using CML cloud runner:

- Easy to install or use. Working the same way for both vendors.
- Maximise the use of your resources using the CI/CD.
- Reproducibility of your ML experiments. Resolving dependencies issues and
  tainted environments.

## Hands on GPU Self-hosted runners 101

### 1) Install nvidia drivers and nvidia-docker in your machine (ubuntu 18.04)

> ```sh

Reviewer 3: let's use `dvc` and start each line with `$ `

Author: @shcheklein sorry I don't understand what do you mean with ```dvc```

Reviewer 3: ah, sorry

I mean that we have a custom highlighter that work better with console. You can enable it by changing `sh` to `dvc`. (it should be renamed into `console` or something eventually).

`RUNNER_REPO` is the url of your Gitlab or Github repo. repo_token is the
personal token generated for your Github or Gitlab repo. Note that for Github
you must check `workflow` along with `repo`.

If everything went fine we should see a runner registered in our repo correctly.

![](/uploads/images/2020-08-05/registered-cml-runner-github.png)

![](/uploads/images/2020-08-05/registered-cml-runner-gitlab.png)

### 3) Setup your Github Actions or Gitlab workflow yaml file to use the runner and commit your changes.

Gitlab

```yaml
train:
  tags:
    - cml
    - gpu

  script:
    - echo 'Hi from CML!' >> report.md
    - cml-send-comment report.md
```

Github

```yaml
name: train-my-model

on: [push]

jobs:
  train:
    runs-on: [self-hosted, cml, gpu]

    steps:
      - uses: actions/checkout@v2

      - name: cml_run
        run: |
          echo 'Hi from CML!' >> report.md
          cml-send-comment report.md
```

# Limitations and future directions

There are still some limitations to be solved at this stage: Github Actions
[can’t run a workflow longer than 72 hours](https://docs.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits).

> Self-hosted runners don’t behave well when they disconnect from the repo,

Reviewer 3: do we have tickets? include links
